src/cocotb/regression.py

```diff
@@ -903,33 +997,27 @@ def generate_tests(
                 else:
                     doc += f"\t{optname}: {repr(optvalue)}\n"
 
-            self.log.info(
-                'Adding generated test "%s" to module "%s"', name, mod.__name__
-            )
-            kwargs = {}
+            kwargs: Dict[str, Any] = {}
             kwargs.update(self.kwargs_constant)
             kwargs.update(testoptions_split)
-            if hasattr(mod, name):
-                self.log.error(
-                    "Overwriting %s in module %s. "
-                    "This causes a previously defined testcase "
-                    "not to be run. Consider setting/changing "
-                    "name_postfix",
-                    name,
-                    mod,
-                )
-            setattr(
-                mod,
-                name,
-                _create_test(
-                    self.test_function,
-                    name,
-                    doc,
-                    mod,
-                    test_dec,
-                    *self.args,
-                    **kwargs,
-                ),
+
+            @functools.wraps(self.test_function)
+            async def _my_test(dut, kwargs: Dict[str, Any] = kwargs):
+                await self.test_function(dut, *self.args, **kwargs)
+
+            _my_test.__doc__ = doc
+            _my_test.__name__ = name
+            _my_test.__qualname__ = name
+
+            cocotb.regression_manager.register_test(
```

Comment:
I think we can solve the over-zealous test importer by filtering the tests on the __module__ attr. The only problem here is if generate_tests() is called on a function that isn't in the current module. We would need to solve this by passing in the module here from the calling location of generate_tests

Reply:
`TestFactory.generate_tests` looking up the stack to find the module to insert tests into is one of the most poorly thought out things I've seen. I'm glad it's dead. `TestFactory`'s usage should be supplanted by `cocotb.parametrize` as it leaves less to interpretation.